thoughts-of-a-rustacean-learning-go.md

```diff
@@ -0,0 +1,244 @@
+explicit; similar to what Rust's [`Sender`][sender] and [`Receiver`][receiver] do in their documentation.
+
+
+[steve]: https://github.com/steveklabnik
+[bufio]: http://golang.org/pkg/bufio/
+[sender]: http://doc.rust-lang.org/std/sync/mpsc/struct.Sender.html
+[receiver]: http://doc.rust-lang.org/std/sync/mpsc/struct.Receiver.html
+
+### Generics
+
+This is a rather common gripe -- Go doesn't have any generics aside from its builtins
+(`chan`s, arrays, slices, and maps can be strongly typed).
+Like my other points about enums and interfaces, we lose out on the ability for advanced type safety here.
+
+Overall it seems like Go doesn't really aim for type safe abstractions,
+preferring runtime matching of types. That's a valid choice to make, though from a Rust background
+I'm not so fond of it.
+
+### Visibility
+
+Visibility (public/private) is done via the capitalization of the field, method, or type name.
+This sort of restriction doesn't hinder usability, but it's quite annoying. The other day I had to make a bunch of fields public to
+satisfy Go's encoding package, and code needed to be updated everywhere with a manual find/replace (the same string was used in other
+contexts too so it couldn't be done automatically)
+
+On the other hand, Rust has a keyword for exporting things, and whilst it has style recommendations
+for the capitalization of variable names (for good reason -- you don't want to accidentally replace an
+enum variant with a wildcard-esque binding in a match, for example), it doesn't error on them or change the
+semantics in any way, just emits a warning. On the other hand, in Go the item suddenly becomes private.
+
+
+### Warnings
+
+Go throws hard errors on finding unused imports or unused variables. This rules out some workflows where
+I partially finish the code and check if it compiles. Quite often whilst debugging I have to add and remove code which uses
+packages like `fmt` or `log`, and I have to scroll up and edit the imports block every time. In Rust, while you _can_ make
+these hard errors, by default they are just warnings.
+
+This isn't really an issue with the language, just a minor annoyance in the implementation.
+
+
+### Conclusion
+
+A major recurring point is that Go seems to advocate runtime over compile time checking (despite being a compiled language),
+something which is totally opposite to what Rust does. This is not just visible in language features
+(like the GC), but also in the tools that are provided — as mentioned above, Go does not give good tools for
+creating type safe abstractions, and the programmer must add dynamic matching over types to overcome this.
+This is similar (though not the same) as what languages like Python and Javascript advocate,
+however these are generally interpreted, not compiled (and come with the benefits of being interpreted),
+so there's a good tradeoff.
+
+Go isn't a language I intend to use for personal projects in the near future. I enjoyed it,
```

Comment:
I think this conclusion might come across as a bit too combative, I might word it a bit more like:

> I wasn't an immediate convert to Go, having said this I really enjoyed learning, and playing with the language. I'm coming from a different paradigm and it would take me time to adjust. Having said this I highly recommend trying out the language.